domain/model/state/state.go

```diff
@@ -32,20 +32,220 @@ func NewState(
+// satisfying modelerrors.AlreadyExists will be returned.
+// If the model type is not found then an error satisfying errors.NotSupported
+// will be returned.
+func Create(
+	ctx context.Context,
+	uuid model.UUID,
+	input model.ModelCreationArgs,
+	tx *sql.Tx,
+) error {
+	if err := createModel(ctx, uuid, tx); err != nil {
+		return err
+	}
+
+	if err := createModelMetadata(ctx, uuid, input, tx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createModel is responsible for establishing the existence of a new model
+// without any associated metadata. If a model with the supplied UUID already
+// exists then an error that satisfies modelerrors.AlreadyExists is returned.
+func createModel(ctx context.Context, uuid model.UUID, tx *sql.Tx) error {
+	stmt := "INSERT INTO model_list (uuid) VALUES (?);"
+	result, err := tx.ExecContext(ctx, stmt, uuid)
+	if jujudb.IsErrConstraintPrimaryKey(err) {
+		return fmt.Errorf("%w for uuid %q", modelerrors.AlreadyExists, uuid)
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+
+	if num, err := result.RowsAffected(); err != nil {
+		return errors.Trace(err)
+	} else if num != 1 {
+		return errors.Errorf("expected 1 row to be inserted, got %d", num)
+	}
+	return nil
+}
+
+// createModelMetadata is responsible for creating a new model metadata record
+// for the given model UUID. If a model metadata record already exists for the
+// given model uuid then an error satisfying modelerrors.AlreadyExists is
+// returned. Conversely should the owner already have a model that exists with
+// the provided name then a modelerrors.AlreadyExists error will be returned. If
+// the model type supplied is not found then a errors.NotSupported error is
 // returned.
-func (s *State) SetCloudCredential(
+//
+// Should the provided cloud and region not be found an error matching
+// errors.NotFound will be returned.
+// If the ModelCreationArgs contains a non zero value cloud credential this func
+// will also attempt to set the model cloud credential using updateCredential. In
+// this  scenario the errors from updateCredential are also possible.
+func createModelMetadata(
 	ctx context.Context,
 	uuid model.UUID,
-	id credential.ID,
+	input model.ModelCreationArgs,
+	tx *sql.Tx,
+) error {
+	cloudRegionUUIDStmt := `
+SELECT cloud_region.uuid
+FROM cloud_region
+INNER JOIN cloud
+ON cloud_region.cloud_uuid = cloud.uuid
+WHERE cloud.name = ?
+AND cloud_region.name = ?
+`
+
+	var (
+		cloudRegionUUID string
+	)
+	err := tx.QueryRowContext(ctx, cloudRegionUUIDStmt, input.Cloud, input.CloudRegion).Scan(&cloudRegionUUID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return fmt.Errorf(
+			"%w cloud region %q for cloud %q",
+			errors.NotFound, input.CloudRegion, input.Cloud,
+		)
+	} else if err != nil {
+		return fmt.Errorf(
+			"getting cloud region uuid for %q and cloud %q: %w",
+			input.CloudRegion, input.Cloud, err,
+		)
+	}
+
+	modelTypeStmt := "SELECT id from model_type WHERE type = ?"
```

Comment:
This is pretty easy to do in the single statement. Then you don't have to pull anything off the wire and into a result set.

Have a look at `ExtendLease` in the lease state.

Reply:
Thanks, I had it this way and perfectly happy to do it this way. The reason I opted over to this path was so that we can produce a better error message. But in hind sigh the validation of the creation args will catch it.